src/GraphService.ts

```diff
@@ -133,42 +133,58 @@ export module GraphService {
 
 
     async function processBatch() {
```

Comment:
Having 2 concerns here:
1. Could be done easier.
2. Missing tests on that. That's a really critical and complex piece, which better be tested.

On the easier remark (and it's not that your code is no working, but if something can be done easier, then it will be easier to maintain in the future):
Idea is simply to check if we've hit 20, then just schedule another processBatch. With code:

```
func processBatch() {
  // early exit if queue is empty
...
  for (const id of Object.keys(processingQueue)) {
    // if we've hit 20 - break
    ...
  }

 // schedule another processBatch(). See how it's done in enqueueRequest()
}
  
}
```

Reply:
Got it! I have a simplification and will add unit tests to exercise the change.